Add/cart lineitems content

haszari opened the pull request:
Fixes #1297 

This PR includes initial implementation of shopping cart line items in the shopping cart - product name, image, quantity UI, line total price etc.

What's implemented in this PR:

- sample (hard-coded) data to show in "full cart" view
- cart title & item count heading
- cart line items (first cut)
  - product image
  - product name
  - quantity selector
  - total price
  - original price (if item is discounted)
- responsive table layout - the price and quantity are laid out differently on mobile

There's a lot of scope in #1297 so this PR doesn't cover it all - will open up new issues for the remaining work. Also there's more work to be done to style this to match mockups (and ensure compatibility with themes).

Here's a list of follow up work - will make follow up issues once this PR's had some reviews. Feel free to go ahead and make these issues (and link to them from this PR description).

- product details column - short description, product variation/selected options
- product details column - low stock warning badge
- #1521 quantity selector custom component 
- remove item from cart - link (desktop) & icon button (mobile)
- line total "save!" badge (if discounted/coupons)
- ensure cart items API response matches schema
- various accessibility issues (tbd)
- final styling (including mobile)
- review `bundlewatch` file size limits (bumped this up while we have inline placeholder image blob_

We'll also need to hook this up to real APIs to actually implement cart behaviour – so we should add an issue for this (or make sure it's covered by an existing issue). Might be best to implement basic cart behaviour before fleshing out above details (e.g. discount badge, remove item etc).

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

<img width="803" alt="Screen Shot 2020-01-08 at 4 25 03 PM" src="https://user-images.githubusercontent.com/4167300/71948041-a0293f00-3233-11ea-9200-36f72b93dbb0.png">

<img width="682" alt="Screen Shot 2020-01-08 at 4 26 22 PM" src="https://user-images.githubusercontent.com/4167300/71948043-a1f30280-3233-11ea-941e-f8ccb2daf035.png">

### How to test the changes in this Pull Request:

1. Edit a page, add cart block, publish page.
2. View page on front end - test in various browsers & environments.

### Changelog

> Add suggested changelog entry here.

nerrad commented:
Unassigning @haszari to signal it can be adopted by someone else until he gets back (or Rua you can pick back up when you get back if this hasn't been adopted).

haszari commented:
FYI - I rebased this onto latest master.

haszari commented:
I'm back on this! Happy to have a co-pilot on this :) cc @nerrad

haszari commented on assets/js/base/components/cart/quantity-selector/index.js:
```diff
@@ -0,0 +1,43 @@
+	const classes = classNames( 'wc-block-quantity-selector-input', className );
+
+	// For now just use a regular number edit.
+	// TODO Replace this with custom control based on design (follow up PR).
```
Added a follow up issue for this: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1521

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-table.js:
```diff
@@ -0,0 +1,53 @@
+
+const CartProductsTable = ( { items = [] } ) => {
+	const products = items.map( ( lineItem ) => {
+		// TODO we'll need to handle srcset, correct image size
```
Now added: 57150df6dc11b6305bf5050ed44232cea57d58cf

nerrad commented:
Oh and one other thing, there's some ci failures caused by the bundlewatch monitor. It looks like the cart frontend is exceeding the set size. This may be unavoidable but you may need to either review why it's that big in size (might be a bundling issue) or bump up the max size configured for this file (use package json, try to just bump up the size for this specific file, not all frontend files).

haszari commented on assets/js/previews/cart-items.js:
```diff
@@ -0,0 +1,79 @@
+// This closely resembles the data returned from the Store API cart/items endpoint.
+// https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/master/src/RestApi/StoreApi#cart-items-api
+export const previewCartItems = [
+	{
```
Curious - I was using an example response from the API as reference (on current branch), not the schema itself. It had `srcset` etc key in `images` array, and all the `currency_*` keys in `totals` object. I'm guessing the API is returning more useful stuff and the schema needs updating (@mikejolley ?). 

In any case this is hard-coded temporary data, so hopefully not blocking merge.

nerrad replied:
I understand it's hardcoded temporary data, but we should still match the shape of the response because that's what will end up getting used in the components when we wire up the api.  Having this in sync with the api ensures the api response is a drop in replacement when we wire it up (and also helps catch any potential bugs in the meantime).

haszari commented on assets/js/base/components/cart/quantity-selector/index.js:
```diff
@@ -0,0 +1,43 @@
+	// 		<span>{ quantity }</span>
+	// 		<button onClick={ () => onChange( quantity + 1 ) }>+</button>
+	// 	</div>
+	// );
```
Good point, always hard letting go of half-finished commented prototypes :)

haszari commented on assets/js/base/components/cart/quantity-selector/index.js:
```diff
@@ -0,0 +1,43 @@
+/**
```
On the fence here, though I'll move it for now. In #1521 we can revisit this.

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-table-item.js:
```diff
@@ -0,0 +1,91 @@
+	quantity,
+	subtotal,
+	total,
+} ) => {
```
Thanks for the tip – the components are leaner now :) cf370746470677bd16b07af11abccd22955d4d8e

haszari commented on assets/js/base/components/cart/quantity-selector/index.js:
```diff
@@ -0,0 +1,43 @@
+import './style.scss';
+
+const QuantitySelector = ( { className, quantity, onChange } ) => {
+	const classes = classNames( 'wc-block-quantity-selector-input', className );
```
> I wonder if we should follow similar a similar schema we've used for css classes of {selector}__{target} (or something like that).

Still getting my head around BEM and how we're using it, will update classes along the lines you've suggested (note some components have changed name so will take that into account).

> Fwiw, I actually think we should be using wc-component as the prefix for components and wc-block only for things that serve directly as the main block container. But that ship has sailed for now I think (open to discussion on it though).

Seems sensible. I was assuming we were using `wc-block` for our "module" prefix (like namespace). `wc-component` seems like a good prefix for shared components (i.e. stuff in `base/components`; rather than use for components used somewhere in a block (i.e. implementation detail/sub components). Something to discuss, especially in context of sharing components with woo core, wc-admin etc :)

Aljullu replied:
Regarding `wc-component` as a prefix, we should take a project-wide decision regarding this and probably #1454 is a good place to discuss it. But until we reach a consensus, I would stick to `wc-block` for everything as we have been doing until now.

One thing that makes me reluctant to use `wc-component` is that there is a package named `@woocommerce/components` which is part of WC Admin, and that prefix seems better suited for that package than for ours.

nerrad replied:
> One thing that makes me reluctant to use wc-component is that there is a package named @woocommerce/components which is part of WC Admin, and that prefix seems better suited for that package than for ours.

That's actually _exactly_ why I want to use `wc-component` here.  One of the criteria I think would be components that will be shared and eventually even moved into the `@woocommerce/components` package.  Using the css prefix would be a good signal for that.

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-table-item.js:
```diff
@@ -0,0 +1,91 @@
+	};
+
+	return (
+		<tr className="wc-block-cart__table-item">
```
This is gone, we don't need a class here right now (remnant from flexbox approach).

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-table.js:
```diff
@@ -0,0 +1,60 @@
+	} );
+
+	return (
+		<table className="wc-block-cart__table">
```
Here's how I rejigged the css classes, let me know if you have better suggestions :)

This is now `wc-block-cart-items` ('block' level) to roughly match the component name – `CartLineItemsTable`.

Inside the table, there's another 'block' `wc-block-cart-item` matching the row component, `CartLineItemRow`.

Generally within these I'm targeting an element (i.e. block__element from BEM, a dependent component within the block), e.g. `wc-block-cart-items__header` or `wc-block-cart-items__header-product`. I think BEM blocks roughly correspond to components .. 

Happy to tweak these in line with our conventions. Now that the component names are fixed up, should be easier to see any untidy class names :)

nerrad replied:
Naming change works for me 👍

haszari commented:
> It looks like the cart frontend is exceeding the set size.

@nerrad I suspect this is because of the huge inline image in the test data (there are two copies of it, as there are two test line items). I ran a local release build and without the image the js file was ~27KB. 

So for now .. I've bumped the size limit just for this file (🤞🏻). If this works, then we can add a follow up issue to reset/review this when we switch to real API data, or look at changing to a smaller test image.

haszari commented:
Thanks for the reviews @mikejolley @nerrad ! I believe I've addressed all the feedback (and also fixed a Safari issue I missed), please re-review and let me know if there's anything that needs another look.

Note I haven't made all the follow up issues yet - the list in the description should cover it all, I'll make new issues first thing tomorrow (unless someone jumps in first) :)

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-table.js:
```diff
@@ -0,0 +1,53 @@
+					{ __( 'Product', 'woo-gutenberg-products-block' ) }
+				</div>
+				<div className="wc-block-cart__table-header-quantity">
+					{ __( 'Quantity', 'woo-gutenberg-products-block' ) }
```
Switched back to a table now. Was using flex as the elements move around a bit for mobile; however needed mobile-specific elements so it's cleaner as a table.

Aljullu commented on assets/js/base/components/cart/quantity-selector/index.js:
```diff
@@ -0,0 +1,43 @@
+ */
+import './style.scss';
+
+const QuantitySelector = ( { className, quantity, onChange } ) => {
```
I applied similar changes to the `<TextInput>` component in #1523 but after @senadir's review and looking at Gutenberg control components, I'm not 100% sure that's what we want.

As @senadir [said](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1523#discussion_r364723700), setting a default for the `quantity`, the `QuantitySelector` will always be a controlled component. Even if that's the [recommended approach](https://reactjs.org/docs/uncontrolled-components.html) according to React Docs, I'm not sure we want to limit this component to always be used like this in case at some point in the future we allow plugins or other projects to use these components.

Looking at our repo, we have been a bit inconsistent with this. We have some components where we don't set a default value or default onChange (`<SortSelect>`, `<RadioControl>`...), while in some other we do (`<CheckboxList>`, `<DropdownSelector>`...). Should we try to agree on a common solution and always stick to it? Gutenberg doesn't seem to set neither a default value neither a default onChange for its form controls (some examples: [1](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/select-control/index.js#L16), [2](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/range-control/index.js#L18), [3](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/textarea-control/index.js#L11), [4](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/text-control/index.js#L11)).

@haszari I don't think this discussion should block this PR at all, but wanted to bring it here to get your attention and @nerrad's. If you think that's an issue we would like to fix, I would open an issue.

nerrad replied:
I agree we should be consistent and I would like to default to making a low level component like this controlled.  Other use-cases are still possible, they just require a wrapper component that's all. I also think whether it's controlled or not depends on the complexity of the component.

So if we want to be consistent about having low level input type components be controlled, then we should do the following imo:

* don't set a default for the event handlers, but ensure PropTypes defines them as required.
* set default for value.

Regarding the Gutenberg examples you gave, I think they should have a default for the event handler otherwise it's going to error if it's not provided (and they aren't controlled). So I actually wouldn't follow the GB pattern here because imo it's wrong (and if I do work in that component I'd fix it :) )

I agree let's not block this pull here, but let's create an issue to make things consistent for low level input type components. 

Thanks for drawing attention to this Albert!

Aljullu replied:
The only drawback I see from this approach is that in `disabled` form controls it's a legit case not to specify an `onChange` prop. But I guess in those cases consumers can just do `onChange={ () => null }` to avoid the warning.

nerrad replied:
> The only drawback I see from this approach is that in disabled form controls it's a legit case not to specify an onChange prop.

I'm sorry I don't follow. `<Disabled />` disables everything on the observed dom in the rendered children. Are you saying that a parent might "disable" an input by not passing through the `onChange` prop?

Aljullu replied:
I mean something like this:
```JSX
<TextInput
	...
	disabled={ true }
	onChange={ () => null } // if the input is disabled, we don't really need an `onChange` handle
/>
```
Consumers will be forced to pass a function to `onChange` even in disabled or readonly inputs. Not a big deal but wanted to put it on the table.

nerrad replied:
Oh I see, well in most use-cases I would simply see it being something like this:

```js
<TextInput
	...
	disabled={ isDisabled }
	onChange={ onChange } 
/>
```

...as there'd be some variable condition controlling whether the field is disabled or not. No need to modify any other props when switching the `disabled` prop?

Aljullu replied:
Yes, I think in most cases that will not be an issue, so your proposed solution sounds good to me. If the others don't say the opposite, I will create an issue to track implementing it in all form components.

Aljullu replied:
Issue created: #1526.

nerrad commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-line-item-row.js:
```diff
@@ -0,0 +1,90 @@
+		<tr>
+			<td className="wc-block-cart-item__product">
+				<div className="wc-block-cart-item__product-wrapper">
+					<img { ...image } alt={ imageAltText } />
```
I agree with @Aljullu 's point here (and I know I suggested this spreading 🤦‍♂ ) .  Let's just setup the img props individually from the `image` object.

nerrad commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-line-item-row.js:
```diff
@@ -0,0 +1,90 @@
+				<div className="wc-block-cart-item__product-wrapper">
+					<img { ...image } alt={ imageAltText } />
+					<div className="wc-block-cart-item__product-details">
+						<div>{ name }</div>
```
Agreed

nerrad commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-line-items-table.js:
```diff
@@ -0,0 +1,43 @@
+
+const CartLineItemsTable = ( { lineItems = [] } ) => {
+	const products = lineItems.map( ( lineItem ) => (
+		<CartLineItemRow key={ lineItem.id } lineItem={ lineItem } />
```
Agreed.

nerrad commented on assets/js/previews/cart-items.js:
```diff
@@ -0,0 +1,81 @@
+	},
+	{
+		key: '2',
+		id: 1,
```
Agreed

haszari commented:
This is approved but there are a couple of things I'm looking at before merge:

- Heading screen reader issue raised by @mikejolley ("Shopping cart 3 items")
- Product images are too big in the editor (new issue)

If fixing these gets complicated I'll add follow up issues instead.

haszari commented on assets/js/blocks/cart-checkout/cart/full-cart/cart-products-title.js:
```diff
@@ -0,0 +1,28 @@
+	return (
+		<h3>
+			<span>{ title } </span>
+			<span className='wc-block-cart__item-count'>{ itemCountHeading }</span>
```
For now I've added an `aria-label` on the heading with a readable version of the visual layout (i.e. add a dash) 6a24613ac215c57c459e8acdd7c4f474bd1124e3. Tested this with macOS VoiceOver (I don't have a lot of accessibility experience) and it seemed reasonable, though there's likely more work to be done in this area.

I'll add this to the follow up issue about accessibility too.
